beacon-chain/sync/rpc_data_column_sidecars_by_root.go

```diff
@@ -0,0 +1,132 @@
+	}
+
+	for i := range columnIdents {
+		if err := ctx.Err(); err != nil {
+			closeStream(stream, log)
+			return err
+		}
+
+		// Throttle request processing to no more than batchSize/sec.
+		if i != 0 && i%batchSize == 0 && ticker != nil {
+			<-ticker.C
+		}
+		s.rateLimiter.add(stream, 1)
+		root, idx := bytesutil.ToBytes32(columnIdents[i].BlockRoot), columnIdents[i].Index
+		// TODO: Differentiate between blobs and columns for our storage engine
+		sc, err := s.cfg.blobStorage.GetColumn(root, idx)
+		if err != nil {
+			if db.IsNotFound(err) {
+				log.WithError(err).WithFields(logrus.Fields{
+					"root":  fmt.Sprintf("%#x", root),
+					"index": idx,
+				}).Debugf("Peer requested data column sidecar by root not found in db")
+				continue
+			}
+			log.WithError(err).Errorf("unexpected db error retrieving data column, root=%x, index=%d", root, idx)
+			s.writeErrorResponseToStream(responseCodeServerError, types.ErrGeneric.Error(), stream)
+			return err
+		}
+
+		// If any root in the request content references a block earlier than minimum_request_epoch,
+		// peers MAY respond with error code 3: ResourceUnavailable or not include the data column in the response.
+		// note: we are deviating from the spec to allow requests for data column that are before minimum_request_epoch,
+		// up to the beginning of the retention period.
+		if sc.SignedBlockHeader.Header.Slot < minReqSlot {
+			s.writeErrorResponseToStream(responseCodeResourceUnavailable, types.ErrDataColumnLTMinRequest.Error(), stream)
+			log.WithError(types.ErrDataColumnLTMinRequest).
+				Debugf("requested data column for block %#x before minimum_request_epoch", columnIdents[i].BlockRoot)
+			return types.ErrDataColumnLTMinRequest
+		}
+
+		SetStreamWriteDeadline(stream, defaultWriteDuration)
+		if chunkErr := WriteDataColumnSidecarChunk(stream, s.cfg.chain, s.cfg.p2p.Encoding(), sc); chunkErr != nil {
+			log.WithError(chunkErr).Debug("Could not send a chunked response")
+			s.writeErrorResponseToStream(responseCodeServerError, types.ErrGeneric.Error(), stream)
+			tracing.AnnotateError(span, chunkErr)
+			return chunkErr
+		}
+	}
+	closeStream(stream, log)
+	return nil
+}
+
+func validateDataColummnsByRootRequest(colIdents types.BlobSidecarsByRootReq) error {
+	if uint64(len(colIdents)) > params.BeaconConfig().MaxRequestDataColumnSidecars {
+		return types.ErrMaxDataColumnReqExceeded
+	}
+	return nil
+}
+
+func DataColumnsRPCMinValidSlot(current primitives.Slot) (primitives.Slot, error) {
+	// Avoid overflow if we're running on a config where deneb is set to far future epoch.
+	if params.BeaconConfig().DenebForkEpoch == math.MaxUint64 || !features.Get().EnablePeerDAS {
+		return primitives.Slot(math.MaxUint64), nil
+	}
+	minReqEpochs := params.BeaconConfig().MinEpochsForDataColumnSidecarsRequest
+	currEpoch := slots.ToEpoch(current)
+	minStart := params.BeaconConfig().DenebForkEpoch
+	if currEpoch > minReqEpochs && currEpoch-minReqEpochs > minStart {
```

Comment:
`>=` instead of `>`?

Reply:
It is already set to `minStart` , we would only want to modify it in the event `currEpoch-minReqEpochs > minStart` is true